### PR TITLE
Clean up boot process

### PR DESCRIPTION
**Pull request by wycats**
This refactor improves the clarity and explicitness of the Application boot process. In particular, it takes
advantage of the previous ApplicationInstance refactoring, and explicitly exposes boot phases via promises.

The end-game here is to allow FastBoot to boot up an application but not create an application instance, and allow individual FastBoot requests to create instances on demand via the `visit()` API.

This will also help the testing infrastructure, which also wants to boot an application once for the entire test suite, and create an instance for each test run.

### CODE REVIEW

File: packages/ember-application/lib/system/application.js [hunk 1]
```diff
@@ -256,28 +256,27 @@ var Application = Namespace.extend(DeferredMixin, {
   init: function() {
     this._super();
```
Comment (stefanpenner) [inline packages/ember-application/lib/system/application.js]: `this._super.apply(this, arguments);`

File: packages/ember-application/lib/system/application.js [hunk 2]
```diff
@@ -256,28 +256,27 @@ var Application = Namespace.extend(DeferredMixin, {
...
+    // decremented by the Application's own `initialize` method.
+    this._readinessDeferrals = 1;
+
+    if (true) {
+      // Create subclass of Ember.Router for this Application instance.
+      // This is to ensure that someone reopening `App.Router` does not
+      // tamper with the default `Ember.Router`.
+      // 2.0TODO: Can we move this into a globals-mode-only library?
+      this.Router = Router.extend();
+      this.waitForDOMReady(this.buildDefaultInstance());
```
Comment (stefanpenner) [inline packages/ember-application/lib/system/application.js]: i know several people who want to optionally begin before DOMReady, any thoughts on how to facilitate that officially?

File: packages/ember-application/lib/system/application.js [hunk 3]
```diff
@@ -671,20 +688,24 @@ var Application = Namespace.extend(DeferredMixin, {
...
-      Ember.BOOTED = true;
-    }
+      if (!Ember.testing) {
+        // Eagerly name all classes that are already loaded
+        Ember.Namespace.processAll();
+        Ember.BOOTED = true;
+      }
 
-    this.resolve(this);
+      this.resolve(this);
```
Comment (stefanpenner) [inline packages/ember-application/lib/system/application.js]: is this bootResolve?

File: packages/ember-application/lib/system/application.js (hunk 2 above)
Comment (chadhietala) [inline packages/ember-application/lib/system/application.js]: Would be great to not wait for DOM ready, especially when you [stream a chunked encoded](http://zoompf.com/blog/2012/05/too-chunky) index.html.  We optimize by streaming down JSON snippets to the index.html. Would be great to hide Ember boot underneath the initial connection.

File: packages/ember-application/lib/system/application.js (hunk 1 above)
Comment (tomdale) [inline packages/ember-application/lib/system/application.js]: Hm. This code is not in a change in this commit.

File: packages/ember-application/lib/system/application.js (hunk 2 above)
Comment (tomdale) [inline packages/ember-application/lib/system/application.js]: Yes, this is absolutely one of the goals of the refactor: to give developers fine-grained control over the boot process. While I would prefer to preserve the current DOM ready semantics, our goal is that developers can prevent the default autobooting behavior, and use the same API as the FastBoot harness to control how the app is booted.

File: packages/ember-application/lib/system/application.js (hunk 3 above)
Comment (tomdale) [inline packages/ember-application/lib/system/application.js]: No, this is for the (deprecated) case of `Application.create().then(() => { })`. I don't think we can remove this until 2.0. Do you disagree?

**Comment by tomdale** [comment]
@rwjblue asked for a diagram of this change, thought it might be useful for others.
untitled 3 2015-01-29 11-51-14

File: packages/ember-application/lib/system/application.js [hunk 4]
```diff
@@ -253,31 +254,44 @@ var Application = Namespace.extend(DeferredMixin, {
...
-
     registerLibraries();
     logLibraryVersions();
 
-    this.scheduleInitialize();
+    // Start off the number of deferrals at 1. This will be
+    // decremented by the Application's own `initialize` method.
+    this._readinessDeferrals = 1;
+
+    if (!Ember.FEATURES.isEnabled('ember-application-visit') || this.autoboot) {
```
Comment (rwjblue) [inline packages/ember-application/lib/system/application.js]: The feature flag system cannot handle this properly.

It **must** be precisely:

```javascript
if (Ember.FEATURES.isEnabled('ember-application-visit')) {
  /* stuff when enabled */
} else {
  /* stuff when disabled */
}
```

**Pull request merged**